vaporetto/src/sentence.rs

```diff
@@ -278,964 +318,900 @@ impl Sentence {
     ///
     /// # Examples
     ///
     /// ```
-    /// use vaporetto::{CharacterType, Sentence};
+    /// use vaporetto::Sentence;
     ///
-    /// let s = Sentence::from_raw("A1あエ漢?").unwrap();
-    /// assert_eq!(&[
-    ///     CharacterType::Roman as u8,
-    ///     CharacterType::Digit as u8,
-    ///     CharacterType::Hiragana as u8,
-    ///     CharacterType::Katakana as u8,
-    ///     CharacterType::Kanji as u8,
-    ///     CharacterType::Other as u8,
-    /// ], s.char_types());
+    /// let s = Sentence::from_tokenized("火星/名詞/カセー に 行き/動詞 まし/助動詞/マシ た").unwrap();
+    /// assert_eq!(2, s.n_tags());
     /// ```
-    pub fn char_types(&self) -> &[u8] {
-        &self.char_type
+    #[inline]
+    pub const fn n_tags(&self) -> usize {
+        self.n_tags
     }
 
-    /// Gets a reference to the boundary score information.
-    ///
-    /// # Returns
-    ///
-    /// If the predictor inserted, the boundary score information is returned. Otherwise, None.
-    pub fn boundary_scores(&self) -> &[i32] {
-        &self.boundary_scores
+    #[inline]
+    pub(crate) fn len(&self) -> usize {
+        self.char_types.len()
     }
 
-    /// Gets a character position in the code point unit.
-    ///
-    /// # Returns
-    ///
-    /// A position in the code point unit.
-    ///
-    /// # Errors
-    ///
-    /// `index` must be a valid position.
-    pub fn get_char_pos(&self, index: usize) -> Result<usize> {
-        if index == 0 {
-            Ok(0)
-        } else {
-            match self.str_to_char_pos.get(index) {
-                Some(index) if *index != 0 => Ok(*index),
-                _ => Err(VaporettoError::invalid_argument("index", "invalid index")),
-            }
-        }
+    #[inline]
+    pub(crate) fn set_predictor(&mut self, predictor: &'b Predictor) {
+        self.predictor.replace(predictor);
+    }
+
+    #[inline]
+    pub(crate) fn str_to_char_pos(&self) -> &[usize] {
+        &self.str_to_char_pos
+    }
+
+    #[inline]
+    pub(crate) fn text_substring(&self, start: usize, end: usize) -> &str {
+        &self.text[self.char_to_str_pos[start]..self.char_to_str_pos[end]]
+    }
+
+    #[cfg(test)]
+    pub(crate) fn char_to_str_pos(&self) -> &[usize] {
+        &self.char_to_str_pos
+    }
+}
+
+/// A Token information.
+#[derive(Clone, Copy)]
+pub struct Token<'a, 'b> {
+    sentence: &'a Sentence<'a, 'b>,
+    start: usize,
+    end: usize,
+}
+
+impl<'a, 'b> Token<'a, 'b> {
+    /// Returns the surface of this token.
+    #[inline]
+    pub fn surface(&self) -> &'a str {
+        self.sentence.text_substring(self.start, self.end)
+    }
+
+    /// Returns tags of this token.
+    #[inline]
+    pub fn tags(&self) -> &'a [Option<Cow<'b, str>>] {
+        let start = (self.end - 1) * self.sentence.n_tags();
+        let end = self.end * self.sentence.n_tags();
+        &self.sentence.tags[start..end]
+    }
+
+    /// Returns the start position of this token.
```

Comment:
```suggestion
    /// Returns the start position of this token in characters.
```
Since `Sentence` handles two types of positions (bytes and characters), it would be better to specify the position type.